File: go/algorithms/bead.go
package algorithms

// BeadSort sorts non-negative integers via the gravity model; falls back to CountingSort for negatives.
func BeadSort(arr []int) []int {
    hasNegative := false
    max := 0
    for _, v := range arr {
        if v < 0 {
            hasNegative = true
        }
        if v > max {
            max = v
        }
    }
    if hasNegative {
        return CountingSort(arr)
    }
    if len(arr) == 0 || max == 0 {
        out := make([]int, len(arr))
        copy(out, arr)
        return out
    }
    grid := make([][]bool, len(arr))
    for i, v := range arr {
        grid[i] = make([]bool, max)
        for j := 0; j < v; j++ {
            grid[i][j] = true
        }
    }
    for j := 0; j < max; j++ {
        sum := 0
        for i := 0; i < len(arr); i++ {
            if grid[i][j] {
                sum++
            }
        }
        for i := len(arr) - 1; i >= 0; i-- {
            grid[i][j] = sum > 0
            if sum > 0 {
                sum--
            }
        }
    }
    out := make([]int, len(arr))
    for i := 0; i < len(arr); i++ {
        h := 0
        for j := 0; j < max && grid[i][j]; j++ {
            h++
        }
        out[i] = h
    }
    return out
}

File: rust/src/bead.rs
use crate::counting_sort;

pub fn bead_sort(arr: &[i32]) -> Vec<i32> {
    let mut has_negative = false;
    let mut max = 0;
    for &v in arr.iter() {
        if v < 0 {
            has_negative = true;
        }
        if v > max {
            max = v;
        }
    }
    if has_negative {
        return counting_sort(arr);
    }
    if arr.is_empty() || max == 0 {
        return arr.to_vec();
    }
    let n = arr.len();
    let mut grid = vec![vec![false; max as usize]; n];
    for (i, &v) in arr.iter().enumerate() {
        for j in 0..v as usize {
            grid[i][j] = true;
        }
    }
    for j in 0..max as usize {
        let mut sum = 0usize;
        for i in 0..n {
            if grid[i][j] {
                sum += 1;
            }
        }
        for i in (0..n).rev() {
            grid[i][j] = sum > 0;
            if sum > 0 {
                sum -= 1;
            }
        }
    }
    let mut out = vec![0i32; n];
    for i in 0..n {
        let mut h = 0;
        while h < max as usize && grid[i][h] {
            h += 1;
        }
        out[i] = h as i32;
    }
    out
}
